File: gt7_engineer/telemetry/listener.py
"""Listener UDP telemetrii GT7 z obsluga heartbeatu.

GT7 wysyla telemetrie dopiero po otrzymaniu pojedynczego bajtu 'A' na porcie
33739. Konsola tnie strumien po okolo 100 pakietach, wiec heartbeat trzeba
okresowo ponawiac.
"""

from __future__ import annotations

import socket
from collections.abc import Iterator

from .decoder import FORMATS, decrypt_packet, parse_packet
from .packet import GT7Packet


class GT7Listener:
    def __init__(
        self,
        playstation_ip: str,
        send_port: int = 33739,
        receive_port: int = 33740,
        heartbeat_every: int = 100,
        timeout: float = 5.0,
        packet_format: str = "A",
    ) -> None:
        self.playstation_ip = playstation_ip
        self.send_port = send_port
        self.receive_port = receive_port
        self.heartbeat_every = max(1, heartbeat_every)
        self.timeout = timeout
        # Format telemetrii decyduje o bajcie heartbeatu i rozmiarze pakietu.
        self.packet_format = packet_format if packet_format in FORMATS else "A"
        spec = FORMATS[self.packet_format]
        self.heartbeat: bytes = spec["heartbeat"]
        self.packet_size: int = spec["size"]
        self._sock: socket.socket | None = None
        self._since_heartbeat = 0

    def __enter__(self) -> "GT7Listener":
        self.open()
        return self

    def __exit__(self, *exc) -> None:
        self.close()

    def open(self) -> None:
        sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        sock.bind(("0.0.0.0", self.receive_port))
        sock.settimeout(self.timeout)
        self._sock = sock
        self._send_heartbeat()

    def close(self) -> None:
        if self._sock is not None:
            self._sock.close()
            self._sock = None

    def _send_heartbeat(self) -> None:
        assert self._sock is not None
        self._sock.sendto(self.heartbeat, (self.playstation_ip, self.send_port))
        self._since_heartbeat = 0

    def receive(self) -> GT7Packet | None:
        """Odbiera i parsuje jeden pakiet. Zwraca None przy timeoucie/blednym pakiecie."""
        if self._sock is None:
            raise RuntimeError("Listener nie jest otwarty - uzyj open() lub kontekstu 'with'.")

        if self._since_heartbeat >= self.heartbeat_every:
            self._send_heartbeat()

        try:
            data, _addr = self._sock.recvfrom(4096)
        except socket.timeout:
            # Brak danych - ponow heartbeat, moze konsola przestala wysylac.
            self._send_heartbeat()
            return None

        self._since_heartbeat += 1

        if len(data) < self.packet_size:
            return None
        decrypted = decrypt_packet(data, self.packet_format)
        if decrypted is None:
            return None
        return parse_packet(decrypted, self.packet_format)

    def stream(self) -> Iterator[GT7Packet]:
        """Nieskonczony generator pakietow (pomija None)."""
        while True:
            packet = self.receive()
            if packet is not None:
                yield packet

File: mobile/lib/telemetry/listener.dart
/// Listener UDP telemetrii GT7 z obsluga heartbeatu.
///
/// Port z `gt7_engineer/telemetry/listener.py`. GT7 wysyla telemetrie dopiero po
/// otrzymaniu pojedynczego bajtu formatu ('A'/'B'/'~') na porcie 33739. Konsola
/// tnie strumien po okolo 100 pakietach, wiec heartbeat trzeba okresowo ponawiac.
///
/// Wystawia strumien [packets] z gotowymi obiektami [Gt7Packet]. Dziala na
/// Androidzie i iOS (dart:io RawDatagramSocket). Telefon musi byc w tej samej
/// sieci Wi-Fi co PlayStation.
library;

import 'dart:async';
import 'dart:io';
import 'dart:typed_data';

import 'decoder.dart';
import 'gt7_packet.dart';
import 'telemetry_source.dart';

class Gt7Listener implements TelemetrySource {
  Gt7Listener({
    required this.playstationIp,
    this.sendPort = 33739,
    this.receivePort = 33740,
    this.heartbeatEvery = 100,
    this.packetFormat = 'A',
  }) : _format = formats.containsKey(packetFormat) ? packetFormat : 'A';

  final String playstationIp;
  final int sendPort;
  final int receivePort;
  final int heartbeatEvery;
  final String packetFormat;

  final String _format;
  RawDatagramSocket? _socket;
  int _sinceHeartbeat = 0;

  final StreamController<Gt7Packet> _controller =
      StreamController<Gt7Packet>.broadcast();

  /// Strumien sparsowanych pakietow (bledne/za krotkie sa pomijane).
  @override
  Stream<Gt7Packet> get packets => _controller.stream;

  @override
  bool get isRunning => _socket != null;

  @override
  String get label => 'PS5 $playstationIp (format $_format)';

  /// Otwiera gniazdo, wysyla pierwszy heartbeat i zaczyna nasluch.
  @override
  Future<void> start() async {
    if (_socket != null) return;
    final spec = formats[_format]!;

    final socket = await RawDatagramSocket.bind(
      InternetAddress.anyIPv4,
      receivePort,
      reuseAddress: true,
    );
    _socket = socket;
    _sendHeartbeat();

    socket.listen((event) {
      if (event != RawSocketEvent.read) return;
      final dg = socket.receive();
      if (dg == null) return;
      final data = Uint8List.fromList(dg.data);
      if (data.length < spec.size) return;

      _sinceHeartbeat++;
      if (_sinceHeartbeat >= heartbeatEvery) _sendHeartbeat();

      final decrypted = decryptPacket(data, _format);
      if (decrypted == null) return;
      _controller.add(parsePacket(decrypted, _format));
    });
  }

  void _sendHeartbeat() {
    final socket = _socket;
    if (socket == null) return;
    final spec = formats[_format]!;
    socket.send(
      Uint8List.fromList([spec.heartbeat]),
      InternetAddress(playstationIp),
      sendPort,
    );
    _sinceHeartbeat = 0;
  }

  /// Zamyka gniazdo i strumien.
  @override
  Future<void> stop() async {
    _socket?.close();
    _socket = null;
    await _controller.close();
  }
}
